src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -508,6 +505,54 @@ public int characteristics() {
         }
     }
 
+    private enum BoundPolicy {
+        SLICE {
+            @Override
+            String format(AbstractMemorySegmentImpl segment, long offset, long size, long length) {
+                return String.format(
+                        "Out of bound access on segment %s; attempting to get slice of length %d at offset %d " +
+                                "which is outside the valid range 0 <= offset+length < byteSize (=%d)",
+                        segment, size, offset, length
+                );
+            }
+        },
+        ACCESS {
+            @Override
+            String format(AbstractMemorySegmentImpl segment, long offset, long size, long length) {
+                return String.format(
+                        "Out of bound access on segment %s; attempting to access an element of length %d at offset %d " +
+                                "which is outside the valid range 0 <= offset+length < byteSize (=%d)",
+                        segment, size, offset, length
+                );
+            }
+        };
+
+        private IndexOutOfBoundsException outOfBoundException(AbstractMemorySegmentImpl segment, long offset, long size,
+                                                              long length) {
+            return new IndexOutOfBoundsException(format(segment, offset, size, length));
+        }
+
+        abstract String format(AbstractMemorySegmentImpl segment, long offset, long size, long length);
+    }
+
+    private static final class BoundsCheckHandler implements BiFunction<String, List<Number>, IndexOutOfBoundsException> {
+        final AbstractMemorySegmentImpl segment;
+        final BoundPolicy policy;
```

Comment:
I recommend turning this into a private record. In addition, we should add a comment that we anticipate this record instance creation to be eliminated by escape analysis when segment access is compiled. I wonder if you have run benchmarks to verify this.